File: modules/module_sample/src/main/java/com/example/william/my/module/sample/activity/LiveDataActivity.java
package com.example.william.my.module.sample.activity;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModelProvider;

import com.alibaba.android.arouter.facade.annotation.Route;
import com.example.william.my.bean.data.ArticleDataBean;
import com.example.william.my.core.retrofit.loading.LoadingTip;
import com.example.william.my.core.retrofit.observer.WithLoadingTipObserver;
import com.example.william.my.module.activity.BaseResponseActivity;
import com.example.william.my.module.router.ARouterPath;
import com.example.william.my.module.sample.model.ArticleViewModel;
import com.google.gson.Gson;

/**
 * https://developer.android.google.cn/topic/libraries/architecture/livedata
 * https://developer.android.google.cn/topic/libraries/architecture/viewmodel
 */
@Route(path = ARouterPath.Sample.Sample_LiveData)
public class LiveDataActivity extends BaseResponseActivity implements LoadingTip.LoadingTipListener {

    private LoadingTip mLoadingTip;
    private ArticleViewModel mArticleViewModel;

    @Override
    public void initView() {
        super.initView();

        initViewModel();
    }

    private void initViewModel() {
        mLoadingTip = LoadingTip.addLoadingTipWithTopBar(this);
        mLoadingTip.setOnReloadListener(this);

        mArticleViewModel = new ViewModelProvider(this).get(ArticleViewModel.class);
        mArticleViewModel.getArticle().observe(this, new WithLoadingTipObserver<ArticleDataBean>(mLoadingTip, "getArticle") {
            @Override
            protected void onResponse(@NonNull ArticleDataBean response) {
                showResponse(new Gson().toJson(response));
            }
        });
        mArticleViewModel.request();
    }

    @Override
    public void reload() {
        mArticleViewModel.request();
    }
}

File: modules/module_sample/src/main/kotlin/com/example/william/my/module/sample/activity/LiveDataActivity.kt
package com.example.william.my.module.sample.activity

import android.os.Bundle
import androidx.activity.viewModels
import com.alibaba.android.arouter.facade.annotation.Route
import com.example.william.my.library.base.BaseActivity
import com.example.william.my.module.router.ARouterPath
import com.example.william.my.module.sample.databinding.SampleLayoutResponseBinding
import com.example.william.my.module.sample.model.LiveDataVMFactory
import com.example.william.my.module.sample.model.LiveDataViewModel

/**
 * LiveData
 * https://developer.android.google.cn/topic/libraries/architecture/livedata
 * <p>
 * LiveData对象不应存在于存储库中。
 * LiveData objects should not live in the repository.
 */
@Route(path = ARouterPath.Sample.Sample_LiveData)
class LiveDataActivity : BaseActivity() {

    private val mViewModel: LiveDataViewModel by viewModels {
        LiveDataVMFactory
    }

    private val username = "17778060027"
    private val password = "ww123456"

    lateinit var mBinding: SampleLayoutResponseBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        mBinding = SampleLayoutResponseBinding.inflate(layoutInflater)
        setContentView(mBinding.root)

        observeViewModel()
    }

    private fun observeViewModel() {
        login()
    }

    private fun login() {
        mViewModel.login.observe(this) {
            mBinding.textView.text = it
        }
        mViewModel.login(username, password)
    }

    /**
     * 通过 使用 Flow 流构造方法 -> asLiveData() 直接返回 LiveData
     */
    private fun loginByFlow() {
        mViewModel.loginByFLow(username, password).observe(this) {
            mBinding.textView.text = it
        }
    }

    /**
     * 通过 使用 Coroutine 协程构造方法 -> liveData<> 直接返回 LiveData
     */
    private fun loginByCoroutine() {
        mViewModel.loginByCoroutine(username, password).observe(this) {
            mBinding.textView.text = it
        }
    }
}